lib/nodejs/sandtable.js

```diff
@@ -0,0 +1,380 @@
+    requestObj.response.send( responseData );
+}
+
+function addToCache( url, data ) {
+    if ( data != null) {
+        // reduce size of requestCache to match cache Limit
+        for ( let key in requestCache ) {
+            if ( Object.keys( requestCache ).length < cacheLimit ) {
+                break;
+            }
+            delete requestCache[ key ];
+        }
+        requestCache[url] = data.toString();
+    }
+}
+
+function nextRequest() {
+    currentRequest = null;
+    processRequest();
+}
+
+function processRequest() {
+    if ( requestQueue.length > 0 && currentRequest == null ) {
+        if ( ipaddress.length <= 0 ) {
+            Init();
+        }
+
+        if ( ipaddress.length <= 0 ) {
+            return;
+        }
+
+        // Use Get Status to ensure that the server is ready prior to sending request
+        var host = ipaddress[0] + ":" + sandTablePort;
+        var reqUrl = "http://" + ipaddress[0] + ":" + sandTablePort + '/OverlayService/GetStatus/?params={}';
+
+
+        const options = url.parse( reqUrl ); 
+        const serverRequest = http.get( reqUrl, function( serverResponse ) {
+            let data = [];
+            serverResponse.on('data', chunk => {
+                data.push(chunk)
+            });
+            serverResponse.on('end', () => {
+                // save to server
+                try {
+                    let dataToSave = JSON.parse(data);
+                    if ( dataToSave && ( dataToSave.status === 0 || dataToSave.status === 2 ) ) { //READY
+                        currentRequest = requestQueue.shift();
+                        while( !currentRequest && requestQueue.length > 0 ) {
+                            currentRequest = requestQueue.shift();
+                        }
+                        if ( currentRequest != null ) {
+                            forwardRequestToSandTable( currentRequest );
+                        }
+                    }
+                } catch ( e ) {
+                    console.log( "error parsing json: " + e.message );
+                }
+
+            } );
+        } ).on( "error", (err) => {
+            console.log( "Error: " + err.message );
+        } );
+    }
+}
+
+function forwardRequestToSandTable( request ) {
+    const req = request.request;
+    const res = request.response;
+    const parsedRequest = request.parsedRequest;
+    const params = request.params;
+    const onData = request.onData;
+    if ( ipaddress.length <= 0 ) {
+        Init();
+    }
+
+    if ( ipaddress.length <= 0 ) {
+        res.writeHeader(500, { 
+          'content-type': 'text/plain' 
+        }); 
+        res.end(process.argv.join(' ') + ':\n\nError ' + serverResponse.statusCode + '\n' + stringifiedHeaders); 
+        return;
+    }
+
+    // change request to match sandtable
+    var newHost = ipaddress[0] + ":" + sandTablePort;
+    var reqUrl = "http://" + newHost + req.url;
+    var originalHost = req.headers.host;
+
+    req.pause(); 
```

Comment:
What's the purpose of the pause() call?

Reply:
pause gets called here while the connection to the SandTableServer is established.  It pauses the flow of data back to the requestor until resume gets called after we pipe the request to the SandTableServer

Reply:
Ok, thanks for the explanation.  Could you add a comment explaining that?